Move Shelley Genesis and aeson instances

kderme opened the pull request:
Shelley Genesis comes from ouroboros-consensus and aeson instances from cardano-node
I need this for https://github.com/input-output-hk/ouroboros-network/issues/1660
Depends on https://github.com/input-output-hk/cardano-base/pull/119

I couldn't push a branch to this repository for some reason..

kderme committed:
Move pure Shelley Genesis from ouroboros-consensus

kderme committed:
Move orphan aeson instances from cardano-node

kderme committed:
Update dependencies

nc6 commented on shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/Genesis.hs:
```diff
@@ -0,0 +1,145 @@
+    , sgMaxMajorPV            :: !Natural
+    , sgMaxLovelaceSupply     :: !Word64
+    , sgProtocolParams        :: !SL.PParams
+    , sgGenDelegs             :: !(Map (SL.KeyHash 'SL.Genesis c) (SL.KeyHash 'SL.GenesisDelegate c))
```
This is out of date, it would need to include the VRF hash. See https://github.com/input-output-hk/ouroboros-network/pull/2196

erikd commented:
input-output-hk/cardano-base#119 is merged.

erikd commented:
The JSON instances in the node repository have been updated. You should probably refresh these ones from there.

erikd commented:
@kderme Were you not able to push a branch directly this is repo? Raising a PR from your own clone of the repo makes it siginificantly more difficult for others to work on this.

erikd commented:
I have rebased this and moved some tests from `node`. It is in the branch https://github.com/input-output-hk/cardano-ledger-specs/tree/erikd/shelley-genesis .

kderme commented:
> @kderme Were you not able to push a branch directly this is repo?

Yes I got error about limited rights.

> I have rebased this and moved some tests from node.

I saw these tests, but they seem to depend on `TPraosStandardCrypto`, which is defined at ouroboros-consensus. Are you sure we can move them here?

erikd commented:
You need to get yourself added to the `cardano-sl` GH group.

I was not even able to build the tests. I suspect that either the tests need to be modified to drop the uses of ` TPraosStandardCrypto` or `TPraosStandardCrypto` also needs to be moved to `ledger-specs`.

JaredCorduan commented:
> You need to get yourself added to the `cardano-sl` GH group.

The `cardano-sl` group wasn't actually added to this repo, but I just gave it write access. @kderme you should be able to push branches now.

dcoutts commented on shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/Genesis.hs:
```diff
@@ -0,0 +1,176 @@
+data ShelleyGenesis c = ShelleyGenesis {
+      sgSystemStart           :: !UTCTime
+    , sgNetworkMagic          :: !Word32
+    , sgNetworkId             :: !SL.Network
+    , sgProtocolMagicId       :: !ProtocolMagicId
```
While we're moving things around, can we please merge these two things or at least name them better.

The terms "network magic" and the "protocol magic" are two terms for the same thing.

The `ProtocolMagicId` is the actual network/protocol magic number. Just a number. This is used in the network handshake protocol, and is used in some of the Byron era crypto, and in some Byron addresses.

The `SL.Network` refers to whether the blockchain is a testnet, or a mainnet, since this info is encoded into Shelley addresses (much like some Byron addresses contain the full magic number, but here just a small bit field with an enumeration of testnet vs mainnet1, mainnet2 etc).

These two are related in the sense that the mainnet always uses a fixed network/protocol magic number.

In the Cardano API we use this type to represent both in one go:
```
data Network
  = Mainnet
  | Testnet !NetworkMagic
```

This means the network magic for mainnet is implicit and has to be hard coded, but that is ok. Alternatively to support multiple mainnets we could use:

```
data Network
  = Mainnet !Int !NetworkMagic
  | Testnet !NetworkMagic
```
That `Int` there refers to which mainnet (Shelley addresses support up to 15), plus the corresponding network magic.

dcoutts reviewed (commented):
Comment below about reducing confusion over the network/protocol magic.

nc6 commented on shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/Genesis.hs:
```diff
@@ -0,0 +1,176 @@
+import Data.Time (NominalDiffTime, UTCTime)
+import Data.Word (Word32, Word64)
+import GHC.Generics (Generic)
+
```
No gaps, one big sorted import list.

nc6 reviewed (approved):
Looks fine, but will need rebasing again, sorry!

kderme commented:
Superseded by https://github.com/input-output-hk/cardano-ledger-specs/pull/1534
